src/probability/upcrossing.lean

```diff
@@ -0,0 +1,914 @@
+      upper_crossing_time a b f N (k + 1),
+    { ext x,
+      simp only [upper_crossing_time_succ_eq, ← ih.2, hitting, set.mem_Ici, tsub_le_iff_right],
+      split_ifs with h₁ h₂ h₂,
+      { simp_rw [← sub_le_iff_le_add, hf x] },
+      { simp_rw [set.mem_Ici, ← hf _ _] at h₂,
+        exact false.elim (h₂ h₁) },
+      { simp_rw [set.mem_Ici, hf _ _] at h₁,
+        exact false.elim (h₁ h₂) },
+      { refl } },
+    refine ⟨this, _⟩,
+    ext x,
+    simp only [lower_crossing_time, this, hitting, set.mem_Iic],
+    split_ifs with h₁ h₂ h₂,
+    { simp_rw [hf' x] },
+    { simp_rw [set.mem_Iic, ← hf' _ _] at h₂,
+      exact false.elim (h₂ h₁) },
+    { simp_rw [set.mem_Iic, hf' _ _] at h₁,
+      exact false.elim (h₁ h₂) },
+    { refl } }
+end
+
+lemma upcrossings_before_pos_eq (hab : a < b) :
+  upcrossings_before 0 (b - a) (λ n x, (f n x - a)⁺) N x = upcrossings_before a b f N x :=
+by simp_rw [upcrossings_before, (crossing_pos_eq hab).1]
+
+private lemma mul_integral_upcrossings_before_le_integral_pos_part'' [is_finite_measure μ]
+  (hf : submartingale f ℱ μ) (hN : 0 < N) (hab : a < b) :
+  (b - a) * μ[upcrossings_before a b f N] ≤ μ[λ x, (f N x - a)⁺] :=
+begin
+  refine le_trans (le_of_eq _) (integral_mul_upcrossings_before_le_integral
+    (hf.sub_martingale (martingale_const _ _ _)).pos
+    (λ x, lattice_ordered_comm_group.pos_nonneg _)
+    (λ x, lattice_ordered_comm_group.pos_nonneg _) hN (sub_pos.2 hab)),
+  simp_rw [sub_zero, ← upcrossings_before_pos_eq hab],
+  refl,
+end
+
+private lemma mul_integral_upcrossings_before_le_integral_pos_part' [is_finite_measure μ]
+  (hf : submartingale f ℱ μ) (hab : a < b) :
+  (b - a) * μ[upcrossings_before a b f N] ≤ μ[λ x, (f N x - a)⁺] :=
+begin
+  by_cases hN : N = 0,
+  { subst hN,
+    simp_rw [← bot_eq_zero, upcrossings_before_bot, bot_eq_zero, integral_const,
+      algebra.id.smul_eq_mul, nat.cast_zero, mul_zero],
+    exact integral_nonneg (λ x, lattice_ordered_comm_group.pos_nonneg _) },
+  { exact mul_integral_upcrossings_before_le_integral_pos_part'' hf (zero_lt_iff.2 hN) hab }
+end
+
+/-- **Doob's upcrossing estimate**: given a real valued discrete submartingale `f` and real
+values `a` and `b`, we have `(b - a) * 𝔼[upcrossings_before a b f N] ≤ 𝔼[(f N - a)⁺]` where
+`upcrossings_before a b f N` is the number of times the process `f` crossed from below `a` to above
+`b` before the time `N`. -/
+lemma submartingale.mul_integral_upcrossings_before_le_integral_pos_part [is_finite_measure μ]
+  (a b : ℝ) (hf : submartingale f ℱ μ) (N : ℕ) :
+  (b - a) * μ[upcrossings_before a b f N] ≤ μ[λ x, (f N x - a)⁺] :=
+begin
+  by_cases hab : a < b,
+  { exact mul_integral_upcrossings_before_le_integral_pos_part' hf hab },
+  { rw [not_lt, ← sub_nonpos] at hab,
+    exact le_trans (mul_nonpos_of_nonpos_of_nonneg hab (integral_nonneg (λ x, nat.cast_nonneg _)))
+      (integral_nonneg (λ x, lattice_ordered_comm_group.pos_nonneg _)) }
+end
+
```

Comment:
You could maybe add a docstring explaining what you are going to do now. Docstrings are very useful to structure a file. If you see other places in this file where this could be helpful, don't hesitate to add other docstrings!